### PR TITLE
Fixed ordering of JVM performance options #8579

### PR DESCRIPTION
**Pull request by jakubmalek**
### Type of change

- Bugfix

### Description

Sorting JVM performance flags with priority to ensure correct ordering.
* To ensure proper ordering I added set of prioritized JVM flags with single `UnlockDiagnosticVMOptions` entry.
* The `-XX` parameters are sorted using priority set to ensure the order.
* If flags are not prioritized they are sorted by name to produce consistent results

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 1]
```diff
@@ -422,27 +413,77 @@ public static void jvmSystemProperties(List<EnvVar> envVars, JvmOptions jvmOptio
...
+        if (isJvmOption1Prioritized && isJvmOption2Prioritized) {
+            return 0;
+        }
+        if (isJvmOption1Prioritized) {
+            return -1;
+        }
+        if (isJvmOption2Prioritized) {
+            return 1;
         }
+        return jvmOption1.getKey().compareTo(jvmOption2.getKey());
```
Comment (jakubmalek) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java:460]: The alphabetical sorting is not really required, I just added it to produce more consistent results, but if you think it's better to try preserve original order (in-case for example `LinkedHashMap` was used) I can remove it.

File: test/src/main/java/io/strimzi/test/annotations/ParallelParameterizedTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * please be sure that you know laws of parallel execution and concurrent programming
+ * be sure that you do not use shared resources, and if you use shared resources please work with synchronization
+ */
+@Target(ElementType.METHOD)
+@Retention(RUNTIME)
+@Execution(ExecutionMode.CONCURRENT)
+@ResourceLock(mode = ResourceAccessMode.READ, value = "global")
+@Documented
+@ParameterizedTest
+public @interface ParallelParameterizedTest {
```
Comment (scholzj) [inline test/src/main/java/io/strimzi/test/annotations/ParallelParameterizedTest.java:32]: Why is this needed? I think we have already other parametrized tests where this was not needed.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java [hunk 3]
```diff
@@ -302,61 +312,166 @@ public void testValidHeapPercentage() {
...
+    void testJvmMemoryOptionsDefaultWithMemoryLimit() {
         Map<String, String> env = heapOptions(jvmOptions(null, "4"), 50, 5_000_000_000L, new ResourceRequirementsBuilder().withLimits(Map.of("memory", new Quantity("1Gi"))).build());
         assertThat(env.get(AbstractModel.ENV_VAR_KAFKA_HEAP_OPTS), is("-Xms4"));
         assertThat(env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_PERCENTAGE), is("50"));
         assertThat(env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_MAX), is("5000000000"));
     }
 
     @ParallelTest
-    public void testJvmMemoryOptionsMemoryRequest() {
+    void testJvmMemoryOptionsMemoryRequest() {
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:355]: Maybe all these changes can be removed from the PR to minimize unnecessary changes and focus on a single task in it?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java [hunk 4]
```diff
@@ -302,61 +312,166 @@ public void testValidHeapPercentage() {
...
     @ParallelTest
-    public void testServiceDnsNames() {
+    void testServiceDnsNames() {
         List<String> dnsNames = ModelUtils.generateAllServiceDnsNames("my-namespace", "my-service");
 
         assertThat(dnsNames.size(), is(4));
         assertThat(dnsNames, hasItems("my-service", "my-service.my-namespace", "my-service.my-namespace.svc", "my-service.my-namespace.svc.cluster.local"));
     }
 
+    @ParallelParameterizedTest
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:370]: Does all this parametrization really add something super special what is worth it? Would it be more useful to just create a private helper method and pass the parameters into it one by one? For Strimzi as a project is not really useful to have 3 unit tests that work in some completely special way. It just makes it harder to maintain the project and do the next contributions.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 5]
```diff
@@ -422,27 +413,77 @@ public static void jvmSystemProperties(List<EnvVar> envVars, JvmOptions jvmOptio
...
+    }
+
+    /**
+     * Parses JVM performance options from {@link JvmOptions#getXx()} property into a space-separated string of JVM flags.
+     * If performance options object is null or XX options are not configured then {@link Optional#empty()} will be returned.
+     *
+     * @param jvmOptions JVM options to parse
+     * @return optional comma-separated string of JVM flags
+     */
+    private static Optional<String> parseJvmPerformanceOptions(JvmOptions jvmOptions) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java:428]: The changes in this class look good to me. But it is quite a lot of various methods. Do you think it would make sense to have some `JvmOptionUtils` class for them or something like that?

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. I left some comments.

**Comment by scholzj** [comment]
Also, could you please try to fix the DCO signoff? The instructions should be under the _Details_ link next to the DCO status.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java (hunk 4 above)
Comment (jakubmalek) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:370]: I think parameterized tests can be used in other places, I generally find them quite useful, but I can remove it.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java (hunk 3 above)
Comment (jakubmalek) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:355]: I'll extract this to proposed J`vmOptionsUtils` class and corresponding `JvmOptionsUtilsTest` and revert changes in here.

File: test/src/main/java/io/strimzi/test/annotations/ParallelParameterizedTest.java (hunk 2 above)
Comment (jakubmalek) [reply test/src/main/java/io/strimzi/test/annotations/ParallelParameterizedTest.java:32]: I can remove it, I just followed the pattern of parallel tests.

Comment (jakubmalek) [reply test/src/main/java/io/strimzi/test/annotations/ParallelParameterizedTest.java]: Removed

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java (hunk 3 above)
Comment (jakubmalek) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java]: I've moved all `JvmOptions` related tests to `JvmOptionsUtilsTest` and restored other changes back to original.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java (hunk 4 above)
Comment (jakubmalek) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java]: Removed

**Review by scholzj — APPROVED** [review]
LGTM. Thanks for the changes.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/JvmOptionUtils.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+     * @return true - JVM option should be prioritized, false - otherwise
+     */
+    private static boolean isPriorityJvmPerformanceOption(Map.Entry<String, String> jvmOption)  {
+        return JVM_PERFORMANCE_PRIORITY_OPTIONS.contains(jvmOption.getKey());
+    }
+
+    /**
+     * Converts JVM option into a JVM flag.
+     *
+     * @param jvmOption JVM option to convert with flag name placed in th eky
```
Comment (fvaleri) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/JvmOptionUtils.java:163]:
```suggestion
     * @param jvmOption JVM option to convert with flag name placed in the key
```

**Review by fvaleri — APPROVED** [review]
LGTM. Tested on Minikube.

```sh
$ kubectl exec my-cluster-kafka-0 -- env | grep JVM_PERFORMANCE
KAFKA_JVM_PERFORMANCE_OPTS=-XX:+UnlockDiagnosticVMOptions -XX:+PrintNMTStatistics
```

**Review by tombentley — APPROVED** [review]
Thanks!

**Comment by scholzj** [comment]
Thanks for the PR @jakubmalek
